src/core/src/type/float8_e4m3.cpp

```diff
@@ -91,27 +99,34 @@ uint8_t f32_to_f8e4m3_bits(const float value) {
             // Use NAN as this type has no infinity
             f8_bits |= (f8e4m3_e_mask | f8e4m3_m_mask);
         } else if (f8_biased_exp > 0) {
-            f8_bits |= (f8_biased_exp << f8e4m3_m_size) | (fractional >> three_bytes_shift);
+            f8_bits |= (f8_biased_exp << f8e4m3_m_size) | (fractional >> byte_shift);
         } else {
             // Restore the hidden 1 in f8 mantissa for subnormal calculation
-            fractional = f8_m_hidden_one_mask | (input & f32_m_mask) << (f32_e_size - f8e4m3_e_size);
-            // Will any bits be shifted off?
-            int32_t shift = f8_biased_exp < -(f8e4m3_e_max) ? 0 : (1U << (1 - f8_biased_exp));
-            uint32_t sticky = (fractional & (shift - 1)) ? 1 : 0;
-
-            fractional = ((1 + f8_biased_exp) > f8e4m3_e_max) ? 0 : fractional >> (1 - f8_biased_exp);
-            fractional |= sticky;
+            fractional = f8_m_hidden_one_mask | (input & f16_m_mask) << (f16_e_size - f8e4m3_e_size);
+            int16_t f8_exp = f8_biased_exp - f8e4m3_e_bias;
+            int16_t shift = 1 - f8_exp;
+            int16_t sticky_mask = f8_exp < f8_e_subnormal_min ? 0 : ((1 << shift) - 1);
+            uint16_t sticky = (fractional & sticky_mask) ? 1 : 0;
+
+            // Subnormal mantissa has less significant bits for smaller exponent
+            fractional = f8_exp < f8_e_subnormal_min ? 0 : fractional >> (1 - f8_biased_exp);
             // apply rounding
-            if (((fractional & round_half) == round_odd) || ((fractional & round_norm) != 0)) {
+            if (((fractional & round_half) == round_odd && sticky == 0) || (fractional & round_norm) != 0 ||
+                sticky != 0) {
                 fractional += round_even;
             }
 
-            f8_bits |= fractional >> three_bytes_shift;
+            f8_bits |= fractional >> byte_shift;
         }
     }
 
     return f8_bits;
 }
+
+uint8_t f32_to_f8e4m3_bits(const float value) {
```

Comment:
If two-step conversion kept maybe this function can be removed and just update
```float8_e4m3::float8_e4m3(const float value) : m_value{f16_to_f8e4m3_bits(static_cast<float16>(value))} {}````.

Reply:
Good point! Applied. Thanks Pawel!